loopy/schedule/tools.py

```diff
@@ -65,6 +74,62 @@ def temporaries_written_in_subkernel(kernel, subkernel):
             for tv in kernel.id_to_insn[insn_id].write_dependency_names()
             if tv in kernel.temporary_variables)
 
+
+def args_read_in_subkernel(kernel, subkernel):
+    from loopy.kernel.tools import get_subkernel_to_insn_id_map
+    insn_ids = get_subkernel_to_insn_id_map(kernel)[subkernel]
+    inames = frozenset().union(*(kernel.insn_inames(insn_id)
+                                 for insn_id in insn_ids))
+    domain_idxs = {kernel.get_home_domain_index(iname) for iname in inames}
+    params = frozenset().union(*(
+        kernel.domains[dom_idx].get_var_names(isl.dim_type.param)
+        for dom_idx in domain_idxs))
+    return (frozenset(arg
+                      for insn_id in insn_ids
+                      for arg in kernel.id_to_insn[insn_id].read_dependency_names()
+                      if arg in kernel.arg_dict)
+            | (params & frozenset(kernel.arg_dict)))
```

Comment:
This'll be more efficient as unioning all the read-deps and then intersecting, maybe?